### ISSUE TITLE
Abstract types are not serialized

### ISSUE THREAD

**Issue by cstockloew** [universAAL/middleware#122]
_Originally Opened: @Alfiva (2012-04-12 19:00:30_)
_Originally Closed: 2013-09-17 15:51:47_

I am not sure that this is actually a bug (and not a design decision) but it is as frustrating.

With the introduction of the new data repr. abstract classes it took me a while realizing that abstract types of a concept are not serialized (sorry if this was explained somewhere). Example:

&quot;apX&quot; is an instance of AssistedPerson.
AssistedPerson is an ont subclass of User, which is an ont subclass of Profilable.
Profilable is an abstract ont class.
When &quot;apX&quot; is serialized its types are listed as:
&quot;apX a ..User, ..AssistedPerson.&quot;
But Profilable is not listed.

I checked that if I make Profilable not abstract it appears.
Could it be possible to include abstract types when serializing?
Or is there some reason to prevent this?

--

From: _this issue has been automatically imported from our old issue tracker_

**Comment by cstockloew** [comment]
_Original author: @cstockloew (2012-04-15 04:12:52)_

Please look at the Javadoc for createNewAbstractOntClassInfo: &quot;An OWL class is abstract iff the Java counterpart is abstract&quot;.

If the Java class is abstract:
call createNewAbstractOntClassInfo
Instances are not serialized; they can not be instantiated

If the Java class is not abstract:
call createNewOntClassInfo
Instances are serialized

Maybe the misunderstanding comes from the fact that OWL has no notion of abstract classes, but in Java it makes sense to have abstract classes, so this case was included. Even if OWL would have abstract classes, why creating an abstract ont class with a non-abstract Java class?

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-04-17 09:58:57)_

I think you got it wrong. I am not complaining about the fact that there are OWL abstract classes, nor that they cannot be instantiated.

I am not instantiating an abstract class. I instantiate a normal class which extends an abstract one. The only thing I want is that this abstract class is listed in the &quot;types&quot; of the instantiated class. Think of the hierarchy:
Profilable &gt; User &gt; AssistedPerson
Where only Profilable is abstract. An instance of AssistedPerson (which is NOT abstract) when serialized will only have the non-abstract types in RDF: User and AssitedPerson. But not the type Profilable. I want it to have all the super-types. Abstract or not: Profilable, User, AssistedPerson.

**Comment by cstockloew** [comment]
_Original author: @cstockloew (2012-04-17 13:27:46)_

OK, I understand. Originally (AFAIK), the serializer was only used for communication between nodes and the de-serialization (and instantiation) on the receiving node. Since abstract types cannot be instantiated, it is just not needed to include this information in the serialization output. This way, it is a bit more efficient (faster and smaller messages). Now, with the additional uses of the serializer this may have changed.

Do you have any specific reason why you want to have this in the turtle output? Something that outweights the efficiency argument?

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-04-17 13:58:02)_

I understand why it was not needed before, but this is in fact a requirement for the additional use of the serializer.
If abstract types are not serialized that means they are not stored and therefore you are losing information that you cannot use later for queries, updates or reasoning. Following the example: if I ask for instances of type Profilable, I will not receive any answer, because no instance will have been received with that type, despite actually being Profilables.

However, there may be a workaround, so don´t change anything yet: It is possible that storage inferencing can add the abstract type if the subclass relationship has been introduced in the store before (like when loading the OWL of the ontology). I have to check that first. However since this feature is not yet implemented, until then the modification of the serializer could be the only option. Stay tuned.

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-04-18 18:58:19)_

Update: I have tested what I introduced in my last post: If the triple
User rdfs:subClassOf Profilable
is introduced in the store (manually or by loading an OWL) the RDFS inferencer realizes the relationship and queries and reasoning will take the additional (abstract) type into account, even if it was not explicitly serialized.

However there is still the issue that components that deserialize such info but don´t rely on the CHE store will not realize the relationship on their own.

This bug will have to remain open until it is decided if it´s better to trade performance for functionality.
